src/routers/StackRouter.js

```diff
@@ -24,6 +24,13 @@ function behavesLikePushAction(action) {
   );
 }
 
+// Gives us a single place to generate a unique
+// key if they didn't provide one
+function addKeyToRoute(key, route) {
+  route.key = key || `Init-${generateKey()}`;
+  return route;
+}
+
```

Comment:
At first I did this in a single place, at the start of `getStateForAction`

```js
getStateForAction(action, state) {
  const key = action.key || `Init-${generateKey()}`;`
  // etc..
```

So that I could things like this:

```js
return {
  routeName: action.routeName,
  params: action.params,
  type: undefined,
  key
},
```

The trouble with that approach was that my testing showed that not all codepaths actually need any key at all, and but in those cases the global `generateKey` counter would still be incremented. Perhaps that's OK, but it made a few tests have unexpected key values (e.g. test writer assumes we're at 0, but we're at 1).

`getStateForAction` is rather complex, so I wanted to touch as little as possible to avoid introducing bugs in codepaths not currently tested.